lib/puppet/provider/package/puppetserver_gem.rb

```diff
@@ -13,47 +17,113 @@
+      rescue Puppet::ExecutionFailure => detail
+        raise Puppet::Error, "Could not list gems: #{detail}", detail.backtrace
+      end
     end
 
+    # When `/tmp` is mounted `noexec`, `puppetserver gem list` will output *** LOCAL GEMS ***
+    #  causing gemsplit to output: Warning: Could not match *** LOCAL GEMS ***
+    gem_list = list
+               .lines
+               .select { |x| x =~ /^(\S+)\s+\((.+)\)/ }
+               .map { |set| gemsplit(set) }
+
     if options[:justme]
-      return list.shift
+      return gem_list.shift
+    else
+      return gem_list
+    end
+  end
+
+  # The puppetserver gem cli commands are particularly slow as they start a JVM.
+  # Instead, for the often-executed list command, use the rubygems library from the puppet ruby,
+  #  setting the GEM_HOME and GEM_PATH to the values in the puppetserver configuration file.
+  # The rubygems library does not have access to java platform gems, for example: json (1.8.3 java),
+  #  but java platform gems should not be managed, by design.
+
+  def self.execute_rubygems_gemlist(gem_regex)
+    pe_puppetserver_conf_file = '/etc/puppetlabs/puppetserver/conf.d/pe-puppet-server.conf'
+    os_puppetserver_conf_file = '/etc/puppetlabs/puppetserver/puppetserver.conf'
+    puppetserver_gem_home = '/opt/puppetlabs/server/data/puppetserver/jruby-gems'
+    puppetserver_gem_path = [puppetserver_gem_home, '/opt/puppetlabs/server/data/puppetserver/vendored-jruby-gems']
+
+    puppetserver_conf_file = Facter.value(:pe_server_version) ? pe_puppetserver_conf_file : os_puppetserver_conf_file
+    puppetserver_conf = Hocon.load(puppetserver_conf_file)
+
+    gem_env = {}
+    if puppetserver_conf.empty?
+      gem_env['GEM_HOME'] = puppetserver_gem_home
+      gem_env['GEM_PATH'] = puppetserver_gem_path
     else
-      return list
+      gem_env['GEM_HOME'] = puppetserver_conf['jruby-puppet']['gem-home']
+      gem_env['GEM_PATH'] = puppetserver_conf['jruby-puppet']['gem-path'].join(':')
     end
+    gem_env['GEM_SPEC_CACHE'] = "/tmp/#{$$}"
+
+    Gem.paths = gem_env
+
+    sio_inn = StringIO.new
+    sio_out = StringIO.new
+    sio_err = StringIO.new
+    stream_ui = Gem::StreamUI.new(sio_inn, sio_out, sio_err, false)
+
+    gem_list_cmd = Gem::Commands::ListCommand.new
+    gem_list_cmd.options[:domain] = :local
+    gem_list_cmd.options[:args] = [gem_regex] if gem_regex
+    gem_list_cmd.ui = stream_ui
+    gem_list_cmd.execute
+
+    # Remove default gems from the list, as the are the default gems from the puppet ruby:
+    #  /opt/puppetlabs/puppet/lib/ruby/gems/x.y.z/specifications/default
+    # There is no method exclude default gems from the local gem list, for example: psych (default: 2.2.2),
+    #  but default gems should not be managed, by design.
+
+    gem_list = sio_out.string.lines.reject { |gem| gem =~ / \(default\: / }
+    gem_list.join("\n")
+  ensure
+    Gem.clear_paths
+    stream_ui.close
+    sio_inn.close
+    sio_out.close
+    sio_err.close
```

Comment:
Closing StringIO just prevents further reads or writes to the object and isn't strictly necessary given these are about to go out of scope and be collected. The `close` method for StreamUI appears to be a no-op as well and is just present to define an interface used by subclasses.

Depending on where an exception gets raised, the values of these variables could be undefined. Instead of adding a `defined?` check to each line, it is probably easier to just skip the `close` calls where resources are not being released.

Reply:
Done.